rust/chains/tw_solana/src/modules/utils.rs

```diff
@@ -75,4 +82,113 @@ impl SolanaTransaction {
+        let instructions = MessageDecompiler::decompile_partly(&tx.message)?;
+        Ok(instructions
+            .iter()
+            .find_map(try_instruction_as_set_unit_price))
+    }
+
+    pub fn get_compute_unit_limit(encoded_tx: &str) -> SigningResult<Option<UnitLimit>> {
+        let tx = VersionedTransaction::from_base64(encoded_tx)?;
+        let instructions = MessageDecompiler::decompile_partly(&tx.message)?;
+        Ok(instructions
+            .iter()
+            .find_map(try_instruction_as_set_unit_limit))
+    }
+
+    pub fn set_compute_unit_price(encoded_tx: &str, price: UnitPrice) -> SigningResult<String> {
+        let tx_bytes = base64::decode(encoded_tx, STANDARD)?;
+        let mut tx: VersionedTransaction =
+            bincode::deserialize(&tx_bytes).map_err(|_| SigningErrorType::Error_input_parse)?;
+        let instructions = MessageDecompiler::decompile_partly(&tx.message)?;
+
+        let set_price_ix = ComputeBudgetInstructionBuilder::set_compute_unit_price(price);
+
+        // First, try to find a `ComputeBudgetInstruction::SetComputeUnitPrice` instruction.
+        let ix_position = instructions
+            .iter()
+            .position(|ix| try_instruction_as_set_unit_price(ix).is_some());
+        // If it presents already, it's enough to update the instruction data only.
+        if let Some(pos) = ix_position {
+            tx.message.instructions_mut()[pos].data = set_price_ix.data;
+            return tx.to_base64().tw_err(|_| SigningErrorType::Error_internal);
+        }
+
+        // `ComputeBudgetInstruction::SetComputeUnitPrice` can be pushed to the end of the instructions list.
+        tx.message
+            .push_simple_instruction(set_price_ix.program_id, set_price_ix.data)?;
+
+        tx.to_base64().tw_err(|_| SigningErrorType::Error_internal)
+    }
+
+    pub fn set_compute_unit_limit(encoded_tx: &str, limit: UnitLimit) -> SigningResult<String> {
+        let tx_bytes = base64::decode(encoded_tx, STANDARD)?;
+        let mut tx: VersionedTransaction =
+            bincode::deserialize(&tx_bytes).map_err(|_| SigningErrorType::Error_input_parse)?;
+        let instructions = MessageDecompiler::decompile_partly(&tx.message)?;
+
+        let set_limit_ix = ComputeBudgetInstructionBuilder::set_compute_unit_limit(limit);
+
+        // First, try to find a `ComputeBudgetInstruction::SetComputeUnitLimit` instruction.
+        let ix_position = instructions
+            .iter()
+            .position(|ix| try_instruction_as_set_unit_limit(ix).is_some());
+        // If it presents already, it's enough to update the instruction data only.
+        if let Some(pos) = ix_position {
+            tx.message.instructions_mut()[pos].data = set_limit_ix.data;
+            return tx.to_base64().tw_err(|_| SigningErrorType::Error_internal);
+        }
+
+        // `ComputeBudgetInstruction::SetComputeUnitLimit` should be at the beginning of the instructions list.
```

Comment:
Hi @satoshiotomakan, I was curious why we should place the SetComputeUnitLimit at the beginning of the instructions list. Could you provide more detailed information?

Reply:
Hi @10gic, I won't try to explain better than this article :)
https://www.helius.dev/blog/priority-fees-understanding-solanas-transaction-fee-mechanics
Jump into the `Best Practices` section

Reply:
Hi @satoshiotomakan, thanks for sharing this link. I suspect that this is outdated information. I found a transaction with 9 instructions (https://explorer.solana.com/tx/2Sca3UiTEaqQ7B9xPsL36gW8zRuBMRK5vDeUGG2Twe8zYfrVz1pwvAzbKHgRERsnDbVKQ6CUGCGJF1BiSQSafCCX), and the SetComputeUnitLimit is the last instruction.

In this transaction, the 7th Instruction consumed more than the default limit (200000):
```
 Program consumed: 243455 of 354569 compute units
```

Reply:
Good to know, thank you for the hint! I think it's fine that we insert the Limit at the beginning, so no need to open a PR to change it

Reply:
Placing SetComputeUnitLimit at the front is also fine.

I noticed that we lack test cases to ensure that the AdvanceNonceAccount instruction is always at the beginning.

I would be happy to provide a PR to move SetComputeUnitLimit to the end of the instructions and update the existing test cases, so we wouldn't need to write new ones to ensure that AdvanceNonceAccount is always at the beginning.

What do you think?